{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e34d5538",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyreadr\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "566bf7b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load expression, genes and samples data\n",
    "expression=pd.read_csv(\"C:/Users/joann/Desktop/M2/Deep_Learning/data/gbm_expression.csv\")\n",
    "genes=pd.read_csv(\"C:/Users/joann/Desktop/M2/Deep_Learning/data/gbm_genes.csv\")\n",
    "samples = pyreadr.read_r(\"C:/Users/joann/Desktop/M2/Deep_Learning/data/gbm_samples.rds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8cd9839a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "odict_keys([None])\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "                                                   barcode       patient  \\\n",
      "rownames                                                                   \n",
      "TCGA-06-6390-01A-11R-A96S-41  TCGA-06-6390-01A-11R-A96S-41  TCGA-06-6390   \n",
      "TCGA-06-5411-01A-01R-1849-01  TCGA-06-5411-01A-01R-1849-01  TCGA-06-5411   \n",
      "TCGA-06-5411-01A-01R-A96S-41  TCGA-06-5411-01A-01R-A96S-41  TCGA-06-5411   \n",
      "TCGA-12-3648-01A-01R-A96T-41  TCGA-12-3648-01A-01R-A96T-41  TCGA-12-3648   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41  TCGA-06-A7TK-01A-21R-A96S-41  TCGA-06-A7TK   \n",
      "\n",
      "                                        sample shortLetterCode  \\\n",
      "rownames                                                         \n",
      "TCGA-06-6390-01A-11R-A96S-41  TCGA-06-6390-01A              TP   \n",
      "TCGA-06-5411-01A-01R-1849-01  TCGA-06-5411-01A              TP   \n",
      "TCGA-06-5411-01A-01R-A96S-41  TCGA-06-5411-01A              TP   \n",
      "TCGA-12-3648-01A-01R-A96T-41  TCGA-12-3648-01A              TP   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41  TCGA-06-A7TK-01A              TP   \n",
      "\n",
      "                                       definition sample_submitter_id  \\\n",
      "rownames                                                                \n",
      "TCGA-06-6390-01A-11R-A96S-41  Primary solid Tumor    TCGA-06-6390-01A   \n",
      "TCGA-06-5411-01A-01R-1849-01  Primary solid Tumor    TCGA-06-5411-01A   \n",
      "TCGA-06-5411-01A-01R-A96S-41  Primary solid Tumor    TCGA-06-5411-01A   \n",
      "TCGA-12-3648-01A-01R-A96T-41  Primary solid Tumor    TCGA-12-3648-01A   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41  Primary solid Tumor    TCGA-06-A7TK-01A   \n",
      "\n",
      "                              intermediate_dimension tumor_descriptor  \\\n",
      "rownames                                                                \n",
      "TCGA-06-6390-01A-11R-A96S-41                     1.0          Primary   \n",
      "TCGA-06-5411-01A-01R-1849-01                     1.1          Primary   \n",
      "TCGA-06-5411-01A-01R-A96S-41                     1.1          Primary   \n",
      "TCGA-12-3648-01A-01R-A96T-41                     0.7          Primary   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                     NaN          Primary   \n",
      "\n",
      "                                                 sample_id  \\\n",
      "rownames                                                     \n",
      "TCGA-06-6390-01A-11R-A96S-41  TCGA-06-6390-01A-11R-A96S-41   \n",
      "TCGA-06-5411-01A-01R-1849-01  TCGA-06-5411-01A-01R-1849-01   \n",
      "TCGA-06-5411-01A-01R-A96S-41  TCGA-06-5411-01A-01R-A96S-41   \n",
      "TCGA-12-3648-01A-01R-A96T-41  TCGA-12-3648-01A-01R-A96T-41   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41  TCGA-06-A7TK-01A-21R-A96S-41   \n",
      "\n",
      "                                             pathology_report_uuid  ...  \\\n",
      "rownames                                                            ...   \n",
      "TCGA-06-6390-01A-11R-A96S-41  dfcad656-40f1-4ecf-a516-bc1fd73d66c1  ...   \n",
      "TCGA-06-5411-01A-01R-1849-01  3c5b40a1-47ca-4a9d-b967-bfc34bdc785c  ...   \n",
      "TCGA-06-5411-01A-01R-A96S-41  3c5b40a1-47ca-4a9d-b967-bfc34bdc785c  ...   \n",
      "TCGA-12-3648-01A-01R-A96T-41  83e094a1-fc5f-4b3b-9f57-827a095bba68  ...   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41  5A1F6CD8-15C3-4D80-8496-340B0A9616D7  ...   \n",
      "\n",
      "                             paper_Transcriptome.Subtype  \\\n",
      "rownames                                                   \n",
      "TCGA-06-6390-01A-11R-A96S-41                          CL   \n",
      "TCGA-06-5411-01A-01R-1849-01                          NE   \n",
      "TCGA-06-5411-01A-01R-A96S-41                          NE   \n",
      "TCGA-12-3648-01A-01R-A96T-41                          CL   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                         NaN   \n",
      "\n",
      "                              paper_Pan.Glioma.RNA.Expression.Cluster  \\\n",
      "rownames                                                                \n",
      "TCGA-06-6390-01A-11R-A96S-41                                     LGr4   \n",
      "TCGA-06-5411-01A-01R-1849-01                                     LGr2   \n",
      "TCGA-06-5411-01A-01R-A96S-41                                     LGr2   \n",
      "TCGA-12-3648-01A-01R-A96T-41                                     LGr4   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                                      NaN   \n",
      "\n",
      "                             paper_IDH.specific.RNA.Expression.Cluster  \\\n",
      "rownames                                                                 \n",
      "TCGA-06-6390-01A-11R-A96S-41                                       NaN   \n",
      "TCGA-06-5411-01A-01R-1849-01                                  IDHwt-R4   \n",
      "TCGA-06-5411-01A-01R-A96S-41                                  IDHwt-R4   \n",
      "TCGA-12-3648-01A-01R-A96T-41                                       NaN   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                                       NaN   \n",
      "\n",
      "                             paper_Pan.Glioma.DNA.Methylation.Cluster  \\\n",
      "rownames                                                                \n",
      "TCGA-06-6390-01A-11R-A96S-41                                     LGm5   \n",
      "TCGA-06-5411-01A-01R-1849-01                                     LGm5   \n",
      "TCGA-06-5411-01A-01R-A96S-41                                     LGm5   \n",
      "TCGA-12-3648-01A-01R-A96T-41                                     LGm4   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                                     LGm4   \n",
      "\n",
      "                              paper_IDH.specific.DNA.Methylation.Cluster  \\\n",
      "rownames                                                                   \n",
      "TCGA-06-6390-01A-11R-A96S-41                                    IDHwt-K2   \n",
      "TCGA-06-5411-01A-01R-1849-01                                    IDHwt-K2   \n",
      "TCGA-06-5411-01A-01R-A96S-41                                    IDHwt-K2   \n",
      "TCGA-12-3648-01A-01R-A96T-41                                    IDHwt-K1   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                                    IDHwt-K1   \n",
      "\n",
      "                             paper_Supervised.DNA.Methylation.Cluster  \\\n",
      "rownames                                                                \n",
      "TCGA-06-6390-01A-11R-A96S-41                         Mesenchymal-like   \n",
      "TCGA-06-5411-01A-01R-1849-01                         Mesenchymal-like   \n",
      "TCGA-06-5411-01A-01R-A96S-41                         Mesenchymal-like   \n",
      "TCGA-12-3648-01A-01R-A96T-41                             Classic-like   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                             Classic-like   \n",
      "\n",
      "                             paper_Random.Forest.Sturm.Cluster  \\\n",
      "rownames                                                         \n",
      "TCGA-06-6390-01A-11R-A96S-41                       Mesenchymal   \n",
      "TCGA-06-5411-01A-01R-1849-01                    RTK I 'PDGFRA'   \n",
      "TCGA-06-5411-01A-01R-A96S-41                    RTK I 'PDGFRA'   \n",
      "TCGA-12-3648-01A-01R-A96T-41                               NaN   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                  RTK II 'Classic'   \n",
      "\n",
      "                              paper_RPPA.cluster  \\\n",
      "rownames                                           \n",
      "TCGA-06-6390-01A-11R-A96S-41                  K1   \n",
      "TCGA-06-5411-01A-01R-1849-01                  K1   \n",
      "TCGA-06-5411-01A-01R-A96S-41                  K1   \n",
      "TCGA-12-3648-01A-01R-A96T-41                 NaN   \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                 NaN   \n",
      "\n",
      "                             paper_Telomere.length.estimate.in.blood.normal..Kb.  \\\n",
      "rownames                                                                           \n",
      "TCGA-06-6390-01A-11R-A96S-41                                                NaN    \n",
      "TCGA-06-5411-01A-01R-1849-01                                             7.3688    \n",
      "TCGA-06-5411-01A-01R-A96S-41                                             7.3688    \n",
      "TCGA-12-3648-01A-01R-A96T-41                                                NaN    \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                                                NaN    \n",
      "\n",
      "                             paper_Telomere.length.estimate.in.tumor..Kb.  \n",
      "rownames                                                                   \n",
      "TCGA-06-6390-01A-11R-A96S-41                                          NaN  \n",
      "TCGA-06-5411-01A-01R-1849-01                                       2.4553  \n",
      "TCGA-06-5411-01A-01R-A96S-41                                       2.4553  \n",
      "TCGA-12-3648-01A-01R-A96T-41                                          NaN  \n",
      "TCGA-06-A7TK-01A-21R-A96S-41                                          NaN  \n",
      "\n",
      "[5 rows x 106 columns]\n"
     ]
    }
   ],
   "source": [
    "print(samples.keys())\n",
    "\n",
    "#extract the dataframe from the dictionary \n",
    "key = next(iter(samples.keys()))\n",
    "samples = samples[key]\n",
    "\n",
    "print(type(samples))\n",
    "print(samples.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "317072da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(391, 106)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['barcode',\n",
       " 'patient',\n",
       " 'sample',\n",
       " 'shortLetterCode',\n",
       " 'definition',\n",
       " 'sample_submitter_id',\n",
       " 'intermediate_dimension',\n",
       " 'tumor_descriptor',\n",
       " 'sample_id',\n",
       " 'pathology_report_uuid',\n",
       " 'submitter_id',\n",
       " 'shortest_dimension',\n",
       " 'sample_type',\n",
       " 'specimen_type',\n",
       " 'longest_dimension',\n",
       " 'days_to_collection',\n",
       " 'state',\n",
       " 'initial_weight',\n",
       " 'tissue_type',\n",
       " 'preservation_method',\n",
       " 'morphology',\n",
       " 'days_to_diagnosis',\n",
       " 'treatments',\n",
       " 'tissue_or_organ_of_origin',\n",
       " 'age_at_diagnosis',\n",
       " 'primary_diagnosis',\n",
       " 'classification_of_tumor',\n",
       " 'tumor_of_origin',\n",
       " 'diagnosis_id',\n",
       " 'site_of_resection_or_biopsy',\n",
       " 'prior_treatment',\n",
       " 'diagnosis_is_primary_disease',\n",
       " 'synchronous_malignancy',\n",
       " 'days_to_last_follow_up',\n",
       " 'prior_malignancy',\n",
       " 'year_of_diagnosis',\n",
       " 'method_of_diagnosis',\n",
       " 'icd_10_code',\n",
       " 'race',\n",
       " 'gender',\n",
       " 'ethnicity',\n",
       " 'vital_status',\n",
       " 'age_at_index',\n",
       " 'days_to_birth',\n",
       " 'demographic_id',\n",
       " 'age_is_obfuscated',\n",
       " 'days_to_death',\n",
       " 'country_of_residence_at_enrollment',\n",
       " 'bcr_patient_barcode',\n",
       " 'primary_site',\n",
       " 'project_id',\n",
       " 'disease_type',\n",
       " 'name',\n",
       " 'releasable',\n",
       " 'released',\n",
       " 'paper_patient',\n",
       " 'paper_Tissue.source.site',\n",
       " 'paper_Study',\n",
       " 'paper_BCR',\n",
       " 'paper_Whole.exome',\n",
       " 'paper_Whole.genome',\n",
       " 'paper_RNAseq',\n",
       " 'paper_SNP6',\n",
       " 'paper_U133a',\n",
       " 'paper_HM450',\n",
       " 'paper_HM27',\n",
       " 'paper_RPPA',\n",
       " 'paper_Histology',\n",
       " 'paper_Grade',\n",
       " 'paper_Age..years.at.diagnosis.',\n",
       " 'paper_Gender',\n",
       " 'paper_Survival..months.',\n",
       " 'paper_Vital.status..1.dead.',\n",
       " 'paper_Karnofsky.Performance.Score',\n",
       " 'paper_Mutation.Count',\n",
       " 'paper_Percent.aneuploidy',\n",
       " 'paper_IDH.status',\n",
       " 'paper_X1p.19q.codeletion',\n",
       " 'paper_IDH.codel.subtype',\n",
       " 'paper_MGMT.promoter.status',\n",
       " 'paper_Chr.7.gain.Chr.10.loss',\n",
       " 'paper_Chr.19.20.co.gain',\n",
       " 'paper_TERT.promoter.status',\n",
       " 'paper_TERT.expression..log2.',\n",
       " 'paper_TERT.expression.status',\n",
       " 'paper_ATRX.status',\n",
       " 'paper_DAXX.status',\n",
       " 'paper_Telomere.Maintenance',\n",
       " 'paper_BRAF.V600E.status',\n",
       " 'paper_BRAF.KIAA1549.fusion',\n",
       " 'paper_ABSOLUTE.purity',\n",
       " 'paper_ABSOLUTE.ploidy',\n",
       " 'paper_ESTIMATE.stromal.score',\n",
       " 'paper_ESTIMATE.immune.score',\n",
       " 'paper_ESTIMATE.combined.score',\n",
       " 'paper_Original.Subtype',\n",
       " 'paper_Transcriptome.Subtype',\n",
       " 'paper_Pan.Glioma.RNA.Expression.Cluster',\n",
       " 'paper_IDH.specific.RNA.Expression.Cluster',\n",
       " 'paper_Pan.Glioma.DNA.Methylation.Cluster',\n",
       " 'paper_IDH.specific.DNA.Methylation.Cluster',\n",
       " 'paper_Supervised.DNA.Methylation.Cluster',\n",
       " 'paper_Random.Forest.Sturm.Cluster',\n",
       " 'paper_RPPA.cluster',\n",
       " 'paper_Telomere.length.estimate.in.blood.normal..Kb.',\n",
       " 'paper_Telomere.length.estimate.in.tumor..Kb.']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#explore the samples metadata\n",
    "print(samples.shape)\n",
    "# list of column names\n",
    "samples.columns.tolist()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "91bc4828",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "barcode                                                  0\n",
      "patient                                                  0\n",
      "sample                                                   0\n",
      "shortLetterCode                                          0\n",
      "definition                                               0\n",
      "                                                      ... \n",
      "paper_Supervised.DNA.Methylation.Cluster               129\n",
      "paper_Random.Forest.Sturm.Cluster                      298\n",
      "paper_RPPA.cluster                                     231\n",
      "paper_Telomere.length.estimate.in.blood.normal..Kb.    325\n",
      "paper_Telomere.length.estimate.in.tumor..Kb.           325\n",
      "Length: 106, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "samples_na=samples.isna().sum()\n",
    "print(samples_na)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ab1377b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "survival_or_time -> ['days_to_collection', 'days_to_diagnosis', 'days_to_last_follow_up', 'vital_status', 'days_to_birth', 'days_to_death', 'paper_Survival..months.', 'paper_Vital.status..1.dead.']\n",
      "demographics -> ['age_at_diagnosis', 'race', 'gender', 'ethnicity', 'age_at_index', 'age_is_obfuscated', 'paper_Age..years.at.diagnosis.', 'paper_Gender']\n",
      "tumor_clinical -> ['tumor_descriptor', 'days_to_diagnosis', 'age_at_diagnosis', 'primary_diagnosis', 'classification_of_tumor', 'tumor_of_origin', 'diagnosis_id', 'diagnosis_is_primary_disease', 'year_of_diagnosis', 'method_of_diagnosis', 'paper_Histology', 'paper_Grade', 'paper_Age..years.at.diagnosis.', 'paper_IDH.status', 'paper_IDH.codel.subtype', 'paper_MGMT.promoter.status', 'paper_Original.Subtype', 'paper_Transcriptome.Subtype', 'paper_IDH.specific.RNA.Expression.Cluster', 'paper_IDH.specific.DNA.Methylation.Cluster', 'paper_Telomere.length.estimate.in.tumor..Kb.']\n",
      "treatment -> ['treatments', 'prior_treatment']\n",
      "technical_batch -> []\n"
     ]
    }
   ],
   "source": [
    "patterns = {\n",
    "    \"survival_or_time\": [\"days\", \"death\", \"follow\", \"surviv\", \"vital\", \"progress\", \"recurr\"],\n",
    "    \"demographics\": [\"age\", \"gender\", \"sex\", \"race\", \"ethnic\"],\n",
    "    \"tumor_clinical\": [\"grade\", \"stage\", \"diagnos\", \"tumor\", \"histolog\", \"subtype\", \"idh\", \"mgmt\"],\n",
    "    \"treatment\": [\"therapy\", \"radiat\", \"chemo\", \"temoz\", \"treatment\", \"drug\"],\n",
    "    \"technical_batch\": [\"center\", \"plate\", \"batch\", \"analyte\", \"aliquot\", \"portion\", \"tss\"],\n",
    "}\n",
    "\n",
    "for group, keys in patterns.items():\n",
    "    hits = [c for c in samples.columns if any(k in c.lower() for k in keys)]\n",
    "    print(group, \"->\", hits[:30])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8759e713",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60660, 392)\n",
      "(60660, 10)\n"
     ]
    }
   ],
   "source": [
    "print(expression.shape)\n",
    "print(genes.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "12662247",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Missing in metadata (expr-only): 1\n",
      "['gene_id']\n",
      "Missing in expression (meta-only): 0\n",
      "[]\n",
      "(60660, 391)\n"
     ]
    }
   ],
   "source": [
    "#compare expression IDs between expression and metadata (diagnose dimension mismatch)\n",
    "expr_ids = expression.columns.astype(str)\n",
    "meta_ids = samples[\"sample_id\"].astype(str)\n",
    "\n",
    "missing_in_meta = sorted(set(expr_ids) - set(meta_ids))\n",
    "missing_in_expr = sorted(set(meta_ids) - set(expr_ids))\n",
    "\n",
    "print(\"Missing in metadata (expr-only):\", len(missing_in_meta))\n",
    "print(missing_in_meta[:20])\n",
    "\n",
    "print(\"Missing in expression (meta-only):\", len(missing_in_expr))\n",
    "print(missing_in_expr[:20])\n",
    "\n",
    "#there are actually 391 samples, but we just treated the gene_id column as a sample column which led to the dimension mismatch\n",
    "#we set gene_id column as index\n",
    "expression = expression.set_index(\"gene_id\")\n",
    "print(expression.shape)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
